File: 200/234 - Palindrome Linked List.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
bool isPalindrome(struct ListNode* head) {
    struct ListNode *slow = head, *fast = head, *temp, *curr, *prev;
    int count = 1;
    
    if (! (head && head->next))
        return 1;
    
    while (fast->next && fast->next->next) {
        slow = slow->next;
        fast = fast->next->next;
        ++count;
    }
    
    if (fast->next)
        curr = slow->next;
    else
        curr = slow;
    prev = NULL;
    
    while (1) {
        temp = curr->next;
        curr->next = prev;
        
        prev = curr;
        if (temp)
            curr = temp;
        else
            break;
    }
    
    while (count--) {
        if (curr->val != head->val)
            return 0;
        curr = curr->next;
        head = head->next;
    }
    
    return 1;
}

File: csharp/234.palindrome-linked-list.cs
/*
 * @lc app=leetcode id=234 lang=csharp
 *
 * [234] Palindrome Linked List
 */
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public bool IsPalindrome(ListNode head) {
        ListNode fast = head, slow = head;
        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }
        fast = fast == null ? slow : slow.next;
        slow = null;
        while (fast != null) {
            ListNode tmp = fast.next;
            fast.next = slow;
            slow = fast;
            fast = tmp;
        }
        while (slow != null) {
            if (slow.val != head.val) {
                return false;
            }
            slow = slow.next;
            head = head.next;
        }
        return true;
    }
}

